### PR TITLE
Add foundations for FIPS flavor

### PR DESCRIPTION
**Pull request by Kaderinho**
### What does this PR do?

This PR adds jobs to create a new Agent package with FIPS support.

### Motivation

Merge foundations to be able to work in parallel on different features 

### Describe how to test/QA your changes

This will be thoroughly tested in future PR where we're gonna
- run all e2e tests
- run specific FIPS tests
- deploy internally (which was done as well for this PR)

### Additional Notes

We don't release the package publicly for now and this work should not impact the main Agent so the temporary lack of testing should not be a problem here.

### CODE REVIEW

**Review by jeremy-hanna — APPROVED** [review]

File: omnibus/config/software/datadog-agent.rb [hunk 1]
```diff
@@ -238,6 +239,38 @@
     delete "#{install_dir}/uselessfile"
   end
 
+  # TODO: move this to omnibus-ruby::health-check.rb
```
Comment (ajgajg1134) [inline omnibus/config/software/datadog-agent.rb:242]: Is this todo being merged intentionally?

**Review by ajgajg1134 — APPROVED** [review]

File: omnibus/config/software/datadog-agent.rb (hunk 1 above)
Comment (jeremy-hanna) [reply omnibus/config/software/datadog-agent.rb:242]: I believe health-check is a post install stage in omnibus and this is a candidate for moving to that? I'm unsure the intent and can remove the comment

**Review by robertjli — APPROVED** [review]

**Review by brycekahle — APPROVED** [review]

**Comment by pducolin** [comment]
Did you run manual test to validate these changes, while waiting for the follow up PR with automated tests?

**Review by pducolin — APPROVED** [review]

File: .gitlab/internal_image_deploy/internal_image_deploy.yml [hunk 2]
```diff
@@ -47,6 +47,51 @@ docker_trigger_internal:
...
 
+docker_trigger_internal-fips:
+  stage: internal_image_deploy
+  rules: !reference [.on_deploy_internal_or_manual]
+  needs:
+    - job: docker_build_fips_agent7
+      artifacts: false
+    - job: docker_build_fips_agent7_arm64
+      artifacts: false
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/deb_x64$DATADOG_AGENT_BUILDIMAGES_SUFFIX:$DATADOG_AGENT_BUILDIMAGES
```
Comment (CelianR) [inline .gitlab/internal_image_deploy/internal_image_deploy.yml:58]:
```suggestion
  image: registry.ddbuild.io/ci/datadog-agent-buildimages/deb_x64$DATADOG_AGENT_BUILDIMAGES_SUFFIX:$DATADOG_AGENT_BUILDIMAGES
```

Now, build images are located to registry.ddbuild.io

**Review by CelianR — APPROVED** [review]
LGTM

File: .gitlab/dev_container_deploy/docker_linux.yml [hunk 3]
```diff
@@ -32,6 +32,35 @@ dev_branch_multiarch-a7:
...
+  extends: .docker_publish_job_definition
+  stage: dev_container_deploy
+  rules: !reference [.manual]
+  needs:
+    - docker_build_dogstatsd_amd64
+    - docker_build_dogstatsd_arm64
+  variables:
+    IMG_REGISTRIES: dev
+    IMG_SOURCES: ${SRC_DSD}:v${CI_PIPELINE_ID}-${CI_COMMIT_SHORT_SHA}-amd64,${SRC_DSD}:v${CI_PIPELINE_ID}-${CI_COMMIT_SHORT_SHA}-arm64
+    IMG_DESTINATIONS: dogstatsd-dev:${CI_COMMIT_REF_SLUG}
```
Comment (alopezz) [inline .gitlab/dev_container_deploy/docker_linux.yml:63]:
```suggestion
```

This looks unrelated and a duplicate of the job defined right after.

File: .gitlab/package_build/linux.yml [hunk 4]
```diff
@@ -80,6 +80,14 @@
   before_script:
     - export RELEASE_VERSION=$RELEASE_VERSION_7
 
+.agent_fips_build:
+  variables:
+    AGENT_MAJOR_VERSION: 7
```
Comment (chouquette) [inline .gitlab/package_build/linux.yml:85]: nit: I don't think you need to specify this, it's already implied by `.agent_7_build`

File: .gitlab/package_build/linux.yml [hunk 5]
```diff
@@ -80,6 +80,14 @@
...
+    FLAVOR: fips
+  before_script:
+    - export RELEASE_VERSION=$RELEASE_VERSION_7
+    - DD_GO_TOOLCHAIN="msgo" source /root/.bashrc
```
Comment (chouquette) [inline .gitlab/package_build/linux.yml:89]: can the same results be achieved by having `DD_GO_TOOLCHAIN` as a regular gitlab variable, and letting `source /root/.bashrc` be sourced by the runner?

What I'm getting at is that I fear we have too many `before_script` to easily reason about them. (that being said we probably could get rid of most since they are mostly about dispatching the agent version between 6 and 7)

File: omnibus/config/software/datadog-agent.rb (hunk 1 above)
Comment (chouquette) [reply omnibus/config/software/datadog-agent.rb:242]: Yes that's the idea, omnibus provides a step that does sanity checking on the built binaries, and it feels like this check would belong in there.
I think the intent from Nicolas was to take a shortcut by introducing the change here rather than having to modify omnibus directly

File: omnibus/config/software/openssl-fips-provider.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+default_version "0.0.1"
+
+OPENSSL_FIPS_MODULE_VERSION="3.0.9"
+OPENSSL_FIPS_MODULE_FILENAME="openssl-#{OPENSSL_FIPS_MODULE_VERSION}.tar.gz"
+OPENSSL_FIPS_MODULE_SHA256_SUM="eb1ab04781474360f77c318ab89d8c5a03abc38e63d65a603cabbf1b00a1dc90"
+
+source url: "https://www.openssl.org/source/#{OPENSSL_FIPS_MODULE_FILENAME}",
+        sha256: "#{OPENSSL_FIPS_MODULE_SHA256_SUM}",
+        extract: :seven_zip,
+        target_filename: "#{OPENSSL_FIPS_MODULE_FILENAME}"
```
Comment (alopezz) [inline omnibus/config/software/openssl-fips-provider.rb:21]:
```suggestion
```

`target_filename` looks unnecessary here: we're not really interested in the tar's name and it's going to default to this anyway.

File: omnibus/config/software/openssl-fips-provider.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    # ---------------- DO NOT MODIFY LINES BELOW HERE ----------------
+    command "./Configure enable-fips"
+
+    command "make"
+    command "make install"
+    # ---------------- DO NOT MODIFY LINES ABOVE HERE ----------------
+
+    mkdir "#{install_dir}/embedded/ssl"
+    mkdir "#{install_dir}/embedded/lib/ossl-modules"
+    copy "/usr/local/lib*/ossl-modules/fips.so", "#{install_dir}/embedded/lib/ossl-modules/fips.so"
```
Comment (alopezz) [inline omnibus/config/software/openssl-fips-provider.rb:38]: Would it make sense to follow the instructions on https://github.com/openssl/openssl/blob/master/README-FIPS.md#copy-the-fips-provider-artifacts-fipsso--fipsmodulecnf-to-known-locations and use `make install_fips` after having configured the environment as for the rest of the software we build? Instead of doing essentially the same thing manually.

File: tasks/libs/common/omnibus.py [hunk 8]
```diff
@@ -113,6 +113,7 @@ def env_filter(item):
             "ENVIRONMENTS",
             "EXPERIMENTS_EVALUATION_ADDRESS",
             "FILTER",
+            "FIPS_MODE",
```
Comment (chouquette) [inline tasks/libs/common/omnibus.py:116]: Since there's an extra software being built when FIPS is enabled, I think we need to account for it when computing the cache key (ie. you should remove this line from the ignore list)

File: tasks/build_tags.py [hunk 9]
```diff
@@ -395,3 +402,10 @@ def compute_config_build_tags(targets="all", build_include=None, build_exclude=N
     build_exclude = [] if build_exclude is None else build_exclude.split(",")
     use_tags = get_build_tags(build_include, build_exclude)
     return use_tags
+
+
+def add_fips_tags(tags: list[str], fips_mode: bool) -> list[str]:
+    if fips_mode:
+        print("Adding goexperiment.systemcrypto build tag for FIPS mode")
```
Comment (chouquette) [inline tasks/build_tags.py:409]: This looks like a leftover debug

File: omnibus/lib/symbols_inspectors.rb [hunk 10]
```diff
@@ -74,6 +78,6 @@ def initialize(binary, &block)
 
   def inspect()
     log.info(self.class.name) { "Inspecting binary #{@binary}" }
-    @block.call(`go tool nm #{@binary}`)
```
Comment (alopezz) [inline omnibus/lib/symbols_inspectors.rb:77]: Why this change?

File: .gitlab/package_build/linux.yml (hunk 5 above)
Comment (alopezz) [inline .gitlab/package_build/linux.yml:89]: What does this do? We could set this variable inside Omnibus where it's specifically needed, given that there's already logic in there to handle FIPS mode; and we don't source `.bashrc` in the other builds as far as I can tell.

**Review by alopezz — CHANGES_REQUESTED** [review]

**Review by L3n41c — APPROVED** [review]
👍🏻 LGTM for the files owned by @DataDog/container-integrations except the duplication of the `dev_branch_multiarch-dogstatsd` job already reported by Alex.

File: .gitlab/package_build/linux.yml (hunk 5 above)
Comment (jeremy-hanna) [reply .gitlab/package_build/linux.yml:89]: > letting `source /root/.bashrc` be sourced by the runner?

do you know at what point /root/.bashrc is sourced by the runner? 

this envvar modifies [the PATH](https://github.com/DataDog/datadog-agent-buildimages/blob/main/setup_go.sh#L37-L43) so if it is exported in the runner at initialization I think that will work but I'm not the most familiar with the timing of this, will try it 👍

File: omnibus/config/software/openssl-fips-provider.rb (hunk 7 above)
Comment (jeremy-hanna) [reply omnibus/config/software/openssl-fips-provider.rb:38]: I don't think so. we have the separate `fipsinstall.sh` to run Step 2 of [the validated 140-2 openssl fips provider](https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4282.pdf) installation instructions. (see appendix A, pg 34)

make install_fips [copies](https://github.com/openssl/openssl/blob/47a80fd2034cd4314d3b4958539dcd3106087109/Configurations/unix-Makefile.tmpl#L667-L679) but does not run the module's self tests or generate the config file output around the self-tests and checksum AFAICT

File: .gitlab/package_build/linux.yml (hunk 5 above)
Comment (jeremy-hanna) [reply .gitlab/package_build/linux.yml:89]: this changes [the PATH](https://github.com/DataDog/datadog-agent-buildimages/blob/main/setup_go.sh#L37-L43) for use the [microsoft/go](https://github.com/microsoft/go/tree/microsoft/main) toolchain for building a fips compliant agent, if we aren't [sourcing bashrc in the runner](https://github.com/DataDog/datadog-agent/pull/31004#discussion_r1850039221) by default then we'll need the before_script to do this

Comment (jeremy-hanna) [reply .gitlab/package_build/linux.yml:89]: https://github.com/DataDog/datadog-agent/pull/31004#discussion_r1850065801

alex suggests we do not source .bashrc by default in runners so I believe this is still necessary

File: omnibus/lib/symbols_inspectors.rb (hunk 10 above)
Comment (jeremy-hanna) [reply omnibus/lib/symbols_inspectors.rb:77]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/9b56d4ede78dac2a6e16330fc01c157c61064699

File: omnibus/config/software/openssl-fips-provider.rb (hunk 6 above)
Comment (jeremy-hanna) [reply omnibus/config/software/openssl-fips-provider.rb]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/9b56d4ede78dac2a6e16330fc01c157c61064699

File: tasks/build_tags.py (hunk 9 above)
Comment (jeremy-hanna) [reply tasks/build_tags.py]: removed in https://github.com/DataDog/datadog-agent/pull/31004/commits/9b56d4ede78dac2a6e16330fc01c157c61064699

File: tasks/libs/common/omnibus.py (hunk 8 above)
Comment (jeremy-hanna) [reply tasks/libs/common/omnibus.py]: removed in https://github.com/DataDog/datadog-agent/pull/31004/commits/9b56d4ede78dac2a6e16330fc01c157c61064699

File: .gitlab/package_build/linux.yml (hunk 4 above)
Comment (jeremy-hanna) [reply .gitlab/package_build/linux.yml]: removed in https://github.com/DataDog/datadog-agent/pull/31004/commits/9b56d4ede78dac2a6e16330fc01c157c61064699

File: .gitlab/internal_image_deploy/internal_image_deploy.yml (hunk 2 above)
Comment (jeremy-hanna) [reply .gitlab/internal_image_deploy/internal_image_deploy.yml]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/4b28e0761ae698629c3f7a21df38fb883e29df51

**Review by safchain — APPROVED** [review]

File: .gitlab/container_build/docker_linux.yml [hunk 11]
```diff
@@ -89,6 +89,33 @@ docker_build_agent7_arm64:
...
+  rules:
+    - !reference [.except_mergequeue]
+    - when: on_success
+  needs:
+    - job: datadog-agent-7-x64-fips
+  variables:
+    IMAGE: registry.ddbuild.io/ci/datadog-agent/agent
+    BUILD_CONTEXT: Dockerfiles/agent
+    TAG_SUFFIX: -7-fips
+    BUILD_ARG: --target test --build-arg PYTHON_VERSION=3 --build-arg DD_AGENT_ARTIFACT=datadog-fips-agent-7*-amd64.tar.xz
```
Comment (chouquette) [inline .gitlab/container_build/docker_linux.yml:104]: Just a note, this is likely to conflict with https://github.com/DataDog/datadog-agent/pull/31483
Nothing critical but better be aware

File: .gitlab/internal_image_deploy/internal_image_deploy.yml [hunk 12]
```diff
@@ -47,6 +47,51 @@ docker_trigger_internal:
       --variable TARGET_ENV
       --variable DYNAMIC_BUILD_RENDER_TARGET_FORWARD_PARAMETERS"
 
+docker_trigger_internal-fips:
```
Comment (chouquette) [inline .gitlab/internal_image_deploy/internal_image_deploy.yml:50]: Not needed for this to be merged, but we could probably share more common code between this job and the regular `docker_trigger_internal` one

File: .gitlab/package_build/linux.yml [hunk 13]
```diff
@@ -83,6 +89,14 @@ datadog-ot-agent-7-x64:
 datadog-ot-agent-7-arm64:
   extends: [.agent_build_common, .agent_build_arm64, .ot_agent_7_build]
 
+# build Agent 7 binaries for x86_64 with FIPS
+datadog-agent-7-x64-fips:
+  extends: [.agent_build_common, .agent_build_x86, .agent_7_build, .agent_fips_build]
```
Comment (chouquette) [inline .gitlab/package_build/linux.yml:94]:
```suggestion
  extends: [.agent_build_common, .agent_build_x86, .agent_fips_build]
```

File: .gitlab/package_build/linux.yml [hunk 14]
```diff
@@ -83,6 +89,14 @@ datadog-ot-agent-7-x64:
...
+
+# build Agent 7 binaries for arm64 with FIPS
+datadog-agent-7-arm64-fips:
+  extends: [.agent_build_common, .agent_build_arm64, .agent_7_build, .agent_fips_build]
```
Comment (chouquette) [inline .gitlab/package_build/linux.yml:98]:
```suggestion
  extends: [.agent_build_common, .agent_build_arm64, .agent_fips_build]
```

File: .gitlab/packaging/rpm.yml [hunk 15]
```diff
@@ -11,7 +11,7 @@
     - printf -- "$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $RPM_GPG_KEY)" | gpg --import --batch
     - EXIT="${PIPESTATUS[0]}"; if [ $EXIT -ne 0 ]; then echo "Unable to locate credentials needs gitlab runner restart"; exit $EXIT; fi
     - RPM_SIGNING_PASSPHRASE=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $RPM_SIGNING_PASSPHRASE) || exit $?; export RPM_SIGNING_PASSPHRASE
-    - inv -e omnibus.build --release-version "$RELEASE_VERSION" --base-dir $OMNIBUS_BASE_DIR --skip-deps --target-project=${DD_PROJECT} ${OMNIBUS_EXTRA_ARGS}
+    - inv -e omnibus.build --release-version "$RELEASE_VERSION" --major-version "$AGENT_MAJOR_VERSION" --base-dir $OMNIBUS_BASE_DIR --skip-deps --target-project=${DD_PROJECT} ${OMNIBUS_EXTRA_ARGS}
```
Comment (chouquette) [inline .gitlab/packaging/rpm.yml:14]: This argument was removed recently, you can omit it
```suggestion
    - inv -e omnibus.build --release-version "$RELEASE_VERSION"  --base-dir $OMNIBUS_BASE_DIR --skip-deps --target-project=${DD_PROJECT} ${OMNIBUS_EXTRA_ARGS}
```

File: .gitlab/packaging/rpm.yml [hunk 16]
```diff
@@ -45,6 +45,7 @@
 .package_rpm_agent_7:
   variables:
     RELEASE_VERSION: $RELEASE_VERSION_7
+    AGENT_MAJOR_VERSION: 7
```
Comment (chouquette) [inline .gitlab/packaging/rpm.yml:48]:
```suggestion
```

File: .gitlab/packaging/rpm.yml [hunk 17]
```diff
@@ -92,6 +93,31 @@ installer_rpm-amd64:
     # explicitly disable the check
     PACKAGE_REQUIRED_FILES_LIST: ""
 
+agent_rpm-x64-a7-fips:
+  extends: [.package_rpm_common, .package_rpm_agent_7, .package_rpm_x86]
+  tags: ["arch:amd64"]
+  needs: ["datadog-agent-7-x64-fips"]
+  variables:
+    DD_PROJECT: agent
```
Comment (chouquette) [inline .gitlab/packaging/rpm.yml:101]: I think you'll need to provide something like `OMNIBUS_EXTRA_ARGS: --flavor=fips`, otherwise the package will have the same name as the regular agent package: https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/721204621#L281

File: omnibus/config/software/datadog-agent.rb [hunk 18]
```diff
@@ -57,6 +59,12 @@
 
   # include embedded path (mostly for `pkg-config` binary)
   env = with_standard_compiler_flags(with_embedded_path(env))
+
+  # Use msgo toolchain when fips mode is enabled
+  if fips_mode? && !windows_target?
+    env["GOROOT"] = msgoroot
+    env["PATH"] = "#{msgoroot}/bin:#{env['PATH']}"
+  end
```
Comment (chouquette) [inline omnibus/config/software/datadog-agent.rb:67]: Non blocking, but I think this would be better suited in `get_omnibus_env` (in `tasks/omnibus.py`)

**Review by chouquette — CHANGES_REQUESTED** [review]

File: .gitlab/package_build/linux.yml (hunk 5 above)
Comment (jeremy-hanna) [reply .gitlab/package_build/linux.yml]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/dada6cbb431637c499d1407f1c43bb5a7b70baaf

Comment (jeremy-hanna) [reply .gitlab/package_build/linux.yml]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/dada6cbb431637c499d1407f1c43bb5a7b70baaf

File: omnibus/config/software/datadog-agent.rb (hunk 18 above)
Comment (jeremy-hanna) [reply omnibus/config/software/datadog-agent.rb:67]: 👍 I will have a follow up PR for this update so we can merge this before the code freeze

File: .gitlab/packaging/rpm.yml (hunk 17 above)
Comment (jeremy-hanna) [reply .gitlab/packaging/rpm.yml:101]: good catch! we have this in deb but not the rpm packaging

Comment (jeremy-hanna) [reply .gitlab/packaging/rpm.yml:101]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/8a52d52c113056e5fad1f3d4bfc4968b66b25d2c

File: .gitlab/packaging/rpm.yml (hunk 16 above)
Comment (jeremy-hanna) [reply .gitlab/packaging/rpm.yml]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/8a52d52c113056e5fad1f3d4bfc4968b66b25d2c

File: .gitlab/packaging/rpm.yml (hunk 15 above)
Comment (jeremy-hanna) [reply .gitlab/packaging/rpm.yml]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/8a52d52c113056e5fad1f3d4bfc4968b66b25d2c

File: .gitlab/package_build/linux.yml (hunk 14 above)
Comment (jeremy-hanna) [reply .gitlab/package_build/linux.yml]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/8a52d52c113056e5fad1f3d4bfc4968b66b25d2c

File: .gitlab/package_build/linux.yml (hunk 13 above)
Comment (jeremy-hanna) [reply .gitlab/package_build/linux.yml]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/8a52d52c113056e5fad1f3d4bfc4968b66b25d2c

File: .gitlab/internal_image_deploy/internal_image_deploy.yml (hunk 12 above)
Comment (jeremy-hanna) [reply .gitlab/internal_image_deploy/internal_image_deploy.yml:50]: Agreed, lets make this change in a follow up PR

File: .gitlab/container_build/docker_linux.yml (hunk 11 above)
Comment (jeremy-hanna) [reply .gitlab/container_build/docker_linux.yml]: fixed in https://github.com/DataDog/datadog-agent/pull/31004/commits/4095a2e65be792f7ee9fd3699a16521d2c75b397

**Review by chouquette — APPROVED** [review]

**Review by alopezz — APPROVED** [review]
